src/emulator/hub.ts

```diff
@@ -0,0 +1,158 @@
+
+  /**
+   * Given a project ID, find and read the Locator file for the emulator hub.
+   * This is useful so that multiple copies of the Firebase CLI can discover
+   * each other.
+   */
+  static readLocatorFile(projectId: string): Locator | undefined {
+    const locatorPath = this.getLocatorFilePath(projectId);
+    if (!fs.existsSync(locatorPath)) {
+      return undefined;
+    }
+
+    const data = fs.readFileSync(locatorPath).toString();
+    const locator = JSON.parse(data) as Locator;
+
+    if (locator.version !== version) {
+      logger.debug(`Found locator with mismatched version, ignoring: ${JSON.stringify(locator)}`);
+      return undefined;
+    }
+
+    return locator;
+  }
+
+  static getLocatorFilePath(projectId: string): string {
+    const dir = os.tmpdir();
+    const filename = `hub-${projectId}.json`;
+    return path.join(dir, filename);
+  }
+
+  private hub: express.Express;
+  private server?: http.Server;
+
+  constructor(private args: EmulatorHubArgs) {
+    this.hub = express();
+    this.hub.use(bodyParser.json());
+
+    this.hub.get("/", async (req, res) => {
+      res.json(this.getLocator());
+    });
+
+    // TODO: Route paths should be constants
+    this.hub.post(EmulatorHub.PATH_EXPORT, async (req, res) => {
+      const exportPath = req.body.path;
+      utils.logLabeledBullet(
+        "emulators",
+        `Received export request. Exporting data to ${exportPath}.`
+      );
+      try {
+        await new HubExport(this.args.projectId, exportPath).exportAll();
+        utils.logLabeledSuccess("emulators", "Export complete.");
+        res.status(200).send("OK");
+      } catch (e) {
+        const errorString = e.message || JSON.stringify(e);
+        utils.logLabeledWarning("emulators", `Export failed: ${errorString}`);
+        res.status(500).send(errorString);
+      }
+    });
+  }
+
+  async start(): Promise<void> {
+    const { host, port } = this.getInfo();
+    this.server = this.hub.listen(port, host);
+    await this.writeLocatorFile();
+  }
+
+  async connect(): Promise<void> {
+    // No-op
+  }
+
+  async stop(): Promise<void> {
+    this.server && this.server.close();
+    await this.deleteLocatorFile();
+  }
+
+  getInfo(): EmulatorInfo {
+    const host = this.args.host || Constants.getDefaultHost(Emulators.HUB);
+    const port = this.args.port || Constants.getDefaultPort(Emulators.HUB);
+
+    return {
+      host,
+      port,
+    };
+  }
+
+  getName(): Emulators {
+    return Emulators.HUB;
+  }
+
+  private getLocator(): Locator {
+    const { host, port } = this.getInfo();
+    const version = pkg.version;
+    return {
+      version,
+      host,
+      port,
+    };
+  }
+
+  private async writeLocatorFile(): Promise<void> {
+    const locatorPath = EmulatorHub.getLocatorFilePath(this.args.projectId);
+    const locator = this.getLocator();
+
+    logger.debug(`[hub] wriing locator at ${locatorPath}`);
+    return new Promise((res, rej) => {
```

Comment:
nit: Rename to `(resolve, reject)` so that people don't have to figure out what `res` means, especially when expressjs stuff is in the same file.